crates/oxc_linter/src/rules/eslint/no_obj_calls.rs

```diff
@@ -0,0 +1,135 @@
+use oxc_ast::{ast::Expression, AstKind};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_semantic::AstNode;
+use oxc_span::{Atom, Span};
+
+use crate::{context::LintContext, rule::Rule};
+
+const NON_CALLABLE_GLOBALS: [&str; 5] = ["Atomics", "Intl", "JSON", "Math", "Reflect"];
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-obj-calls): Disallow calling some global objects as functions")]
+#[diagnostic(severity(error), help("{0} is not a function."))]
+struct NoObjCallsDiagnostic(Atom, #[label] pub Span);
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NoObjCalls;
+
+impl Default for NoObjCalls {
+    fn default() -> Self {
+        Self
+    }
+}
+
+declare_oxc_lint! {
+    /// ### What it does
+    /// Disallow calling some global objects as functions
+    ///
+    /// ### Why is this bad?
+    /// Some global objects are not intended to be called as functions.
+    /// Calling them as functions will usually result in a TypeError being thrown.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// let math = Math();
+    /// let newMath = new Math();
+    ///
+    /// let json = JSON();
+    /// let newJson = new JSON();
+    ///
+    /// let atomics = Atomics();
+    /// let newAtomics = new Atomics();
+    ///
+    /// let intl = Intl();
+    /// let newIntl = new Intl();
+    ///
+    /// let reflect = Reflect();
+    /// let newReflect = new Reflect();
+    ///
+    /// // Good
+    /// let area = r => 2 * Math.PI * r * r;
+    /// let object = JSON.parse("{}");
+    /// let first = Atomics.load(sharedArray, 0);
+    /// let segmenterFrom = Intl.Segmenter("fr", { granularity: "word" });
+    /// ```
+    NoObjCalls,
+    correctness,
+}
+
+fn is_global_obj<'a>(str: impl PartialEq<&'a str>) -> bool {
+    NON_CALLABLE_GLOBALS.iter().any(|&n| str == n)
+}
+
+impl Rule for NoObjCalls {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let (callee, span) = match node.kind() {
+            AstKind::NewExpression(expr) => (&expr.callee, expr.span),
+            AstKind::CallExpression(expr) => (&expr.callee, expr.span),
+            _ => return,
+        };
+
+        let ident: Atom = match callee {
+            // handle new Math(), Math(), etc
+            Expression::Identifier(ident) => {
+                ident.name.clone()
```

Comment:
To avoid an extra allocation, we can return the reference here.

Reply:
Fixed